l1-contracts/contracts/l2-upgrades/L2TestnetSystemProxiesUpgrade.sol

```diff
@@ -0,0 +1,101 @@
+    function _initProxyOnAddress(
+        address _address,
+        bytes memory _fullBytecodeInfo,
+        bytes memory _bytecodeInfoSystemProxy
+    ) internal {
+        L2GenesisForceDeploymentsHelper.unsafeForceDeployZKsyncOS(_bytecodeInfoSystemProxy, _address);
+        ISystemContractProxy(_address).forceInitAdmin(L2_SYSTEM_CONTRACT_PROXY_ADMIN_ADDR);
+
+        L2GenesisForceDeploymentsHelper.updateZKsyncOSContract(_fullBytecodeInfo, _address);
+    }
+
+    /// @notice The function to be delegate-called by the L2ComplexUpgrader.
+    /// @dev It will force deploy system proxies as well as the SystemContractProxyAdmin to ensure
+    /// that system contracts will use those for future upgrades.
+    /// @dev It is assumed that no additional initialization is needed, all the L2 contracts have been properly initialized and
+    /// so it is enough to just:
+    /// 1. force deploy the SystemProxy bytecode on top of the current contract as well as the SystemContractProxyAdmin.
+    /// 2. force deploy the implementations and upgrade proxy implementations.
+    // slither-disable-next-line locked-ether
+    function upgrade(
+        bytes calldata _fixedForceDeploymentsData,
+        bytes calldata _systemContractProxyAdminBytecodeInfo,
+        bytes calldata _complexUpgraderProxyBytecodeInfo
+    ) external {
+        // Decode the fixed and additional force deployments data.
+        FixedForceDeploymentsData memory fixedForceDeploymentsData = abi.decode(
+            _fixedForceDeploymentsData,
+            (FixedForceDeploymentsData)
+        );
+
+        L2GenesisForceDeploymentsHelper.forceDeployOnAddressZKsyncOS(
+            _systemContractProxyAdminBytecodeInfo,
+            L2_SYSTEM_CONTRACT_PROXY_ADMIN_ADDR
+        );
+        SystemContractProxyAdmin(L2_SYSTEM_CONTRACT_PROXY_ADMIN_ADDR).forceSetOwner(address(this));
+
+        // It is expected that all zksync os bytecodes infos contain two parts:
+        // the abi-encoding of the implementation and the actual system proxy.
+        // So we split the two here:
+        (, bytes memory bytecodeInfoSystemProxy) = abi.decode(
+            (fixedForceDeploymentsData.bridgehubBytecodeInfo),
+            (bytes, bytes)
+        );
+
+        _initProxyOnAddress(
+            L2_MESSAGE_ROOT_ADDR,
+            fixedForceDeploymentsData.messageRootBytecodeInfo,
+            bytecodeInfoSystemProxy
+        );
+        _initProxyOnAddress(
+            L2_BRIDGEHUB_ADDR,
+            fixedForceDeploymentsData.bridgehubBytecodeInfo,
+            bytecodeInfoSystemProxy
+        );
+        _initProxyOnAddress(
+            L2_ASSET_ROUTER_ADDR,
+            fixedForceDeploymentsData.l2AssetRouterBytecodeInfo,
+            bytecodeInfoSystemProxy
+        );
+        _initProxyOnAddress(
+            L2_NATIVE_TOKEN_VAULT_ADDR,
+            fixedForceDeploymentsData.l2NtvBytecodeInfo,
+            bytecodeInfoSystemProxy
+        );
+        _initProxyOnAddress(
```

Comment:
is there a reason the v29 interop contracts are not deployed? I can see two arguments, don't deploy until the feature works, VS deploy now and upgrade if needed. 

Not really an issue, just bringing up